### PR TITLE
[rpc] Fix transaction size comments and RPC help text.

### PR DESCRIPTION
**Pull request by jnewbery**
As part of BIP 141, transaction sizes have been replaced by virtual size (with witness data discounted) as the metric for measuring the fee rate of a transaction. RPC now returns the virtual size of the transaction instead of the serialized size. However, code comments and RPC help text have not been updated everywhere to reflect this change.

This PR fixes various comments and RPC help text.

### CODE REVIEW

**Review by sdaftuar — APPROVED** [review]

**Comment by MarcoFalke** [comment]
utACK 2ff29a7

**Review by laanwj — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/8747/commits/2ff29a76766249ddc92ea210b16699dd705e0343

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -331,18 +331,18 @@ UniValue getdifficulty(const UniValue& params, bool fHelp)
 
 std::string EntryDescriptionString()
 {
-    return "    \"size\" : n,             (numeric) transaction size in bytes\n"
+    return "    \"size\" : n,             (numeric) virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted.\n"
```
Comment (luke-jr) [inline src/rpc/blockchain.cpp]: This is a bug. :(

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -1236,7 +1236,7 @@ UniValue getmempoolinfo(const UniValue& params, bool fHelp)
             "\nResult:\n"
             "{\n"
             "  \"size\": xxxxx,               (numeric) Current tx count\n"
-            "  \"bytes\": xxxxx,              (numeric) Sum of all tx sizes\n"
+            "  \"bytes\": xxxxx,              (numeric) Sum of all virtual transaction sizes as defined in BIP 141. Differs from actual serialized size because witness data is discounted\n"
```
Comment (luke-jr) [inline src/rpc/blockchain.cpp]: This also shouldn't be virtual.

**Review by luke-jr — CHANGES_REQUESTED** [review]
The reasonable/expected behaviour in at least these cases is to give real sizes, not "virtual size" nonsense.

**Review by jonasschnelli — APPROVED** [review]

File: src/rpc/blockchain.cpp (hunk 1 above)
Comment (jonnynewbs) [inline src/rpc/blockchain.cpp]: This commit updates the RPC help text to reflect what the function call actually returns. If you think there's a problem with the RPC function, I suggest you raise an issue against that code.

File: src/rpc/blockchain.cpp (hunk 2 above)
Comment (jonnynewbs) [inline src/rpc/blockchain.cpp]: See above

File: src/rpc/blockchain.cpp (hunk 1 above)
Comment (luke-jr) [inline src/rpc/blockchain.cpp]: Well, NACK changing documentation to match buggy behaviour rather than fixing the bug...

**Comment by sipa** [comment]
I disagree this is a bug. Apart from miners, nobody should care about the
total serialized size.

**Comment by laanwj** [comment]
@sipa so do you agree on this documentation change?

**Comment by sipa** [comment]
@laanwj I agree with it, though there was some discussion in one of the former IRC meetings that we shouldn't call this virtual size.

**Comment by luke-jr** [comment]
If we're not showing the size in bytes, then just get rid of the size/bytes fields and add weight...

**Comment by jonnynewbs** [comment]
Removing fields from objects returned by RPC functions is a non-backwards compatible change, which may break clients using the API. I agree that adding weight may be useful for some users and we should consider extending the RPC return objects to include it. If you feel strongly about having weight returned RPC functions, let me know and I'll open a new PR.

This PR is simply to fix up the help text so it reflects what the function call is actually doing.

**Comment by TheBlueMatt** [comment]
Strinctly speaking, this is wrong...BIP 141 defines virtual size as "Transaction weight / 4 (rounded up to nearest integer)." with weight being " Base transaction size * 3 + Total transaction size (ie. the same method as calculating Block weight from Base size and Total size).".

However, this method actually returns that iff sigops per byte of the tx is lower than sigops per byte of the block consensus rule. Not sure if it matters since that should be rare, but it is, strictly speaking, different.

**Comment by jnewbery** [comment]
@TheBlueMatt - you're correct. The actual value returned will be higher than the virtual size if the sigsops per byte in the tx is greater the nBytesPerSigOp due to https://github.com/bitcoin/bitcoin/commit/ab942c15bd3854650afa810d7c22d7fd30d346c1.

I *could* expand the RPC text from:
```(numeric) Sum of all virtual transaction sizes as defined in BIP 141. Differs from actual serialized size because witness data is discounted\n"```
to
```(numeric) Sum of all virtual transaction sizes as defined in BIP 141. Differs from actual serialized size because witness data is discounted. If the sigsops per byte in the transaction is greater nBytesPerSigOp, then returns a size based on the cost of the sigops rather than the serialized size of the transaction.\n"```

I don't think that actually makes things clearer, but I'm happy to update the text if you think that's preferable.

**Comment by TheBlueMatt** [comment]
Its relatively strange to me that it would be inconsistent with the BIP. If we want to keep it this way, I'd say changing the BIP is sufficient as the help text refers to that. Alternatively, we could change the return value to be consistent with the BIP. Either way, I agree, expanding the RPC help text is strange.

**Comment by laanwj** [comment]
Closing this, as there seems to be strong disagreement about this and the discussion has bled out.

**Comment by morcos** [comment]
@laanwj please don't close this.   I don't think anyone (other than possibly @luke-jr) disagrees that this is an clear improvement.  The existing help text is confusing and this is far more accurate.  @TheBlueMatt 's comments are a more obscure technical refinement, but I don't think meant as a blocker to merging this. 

I think its important that we make an effort to properly document and educate our users about how the software works.  The whole concept of "virtual size" is a bit unfortunate but I don't see any way around keeping it unless we want to radically change how people think about feerates.  We can't be held up from making clear improvements to the software just because one person vehemently disagrees.

utACK from me

**Comment by jnewbery** [comment]
Travis appeared to fail because of an unrelated environment setup issue. I've rebased this PR which will kick off another travis run.

**Comment by bitcoinfees** [comment]
There seems to be a conflict - `getrawmempool` defines "size" as the virtual size, whereas `getrawtransaction` defines it as the serialized size (and using "vsize" instead for virtual size).  This doc change helps to clarify the difference, but to avoid confusion perhaps the term "size" should be made to mean the same thing in all RPC calls.

Oh and in `getblock` RPC, "size" also refers to serialized size and not "weight / 4".

**Comment by jnewbery** [comment]
@bitcoinfees - yes, I completely agree with everything you say, and that it's confusing that 'size' means different things in different RPCs.

I didn't make any functional changes to the API since there may be applications that rely on the current behaviour. This PR simply updates the self-documentation in the RPCs so that it accurately describes what is being returned.

Perhaps we should change those RPC definitions, but that's a discussion for another PR and I don't think it should block this documentation fix.

**Comment by luke-jr** [comment]
It still doesn't make sense to document a bug rather than fixing it.

**Comment by morcos** [comment]
@luke-jr but that is not the choice at hand.  The choice presented by this PR is to document the existing behavior or leave it improperly documented.  Choosing to document it does not preclude changing it in the future.  Although I think most of us do not consider it a bug.

I can't understand why this is even a question.  How can we argue against more clearly explaining to users what the behavior is?  We should always be willing to document existing behavior instead of stalling on that while arguing about whether to change the behavior.

**Comment by jnewbery** [comment]
@luke-jr - I agree with @morcos. Nothing in this documentation fix precludes you from opening an issue or PR to change the API.

**Comment by morcos** [comment]
@sipa @TheBlueMatt This PR has several ACK's (and @luke-jr's objection), but I think its held up from merging b/c lack of clear feedback from you two.  If you are OK with it could you say?

**Comment by sipa** [comment]
ACK 2ddb53054579c70de061d95386e9a5a02255d53f

File: src/rpc/mining.cpp [hunk 3]
```diff
@@ -840,7 +841,8 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
             "\nWARNING: This interface is unstable and may disappear or change!\n"
             "\nEstimates the approximate fee per kilobyte needed for a transaction to begin\n"
             "confirmation within nblocks blocks if possible and return the number of blocks\n"
-            "for which the estimate is valid.\n"
+            "for which the estimate is valid. Uses virtual transaction size as defined\n"
+            "in BIP 141 (witness data is discounted)\n"
```
Comment (paveljanik) [inline src/rpc/mining.cpp]: Dot at the end of sentence missing.

**Comment by jnewbery** [comment]
Thanks @paveljanik . Comment fixed.

**Comment by paveljanik** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/8747/commits/d29505db22892548a5eca5567bf8acf4e950efea

**Pull request merged**